### ISSUE TITLE
BPTTIterator: Result of Slicing is an empty tensor

### ISSUE THREAD

**Issue by demiguo** [pytorch/text#137]
I can't iterate through all the batches in the iterator returned by BPTTITerator.
For example, this will give an error: batches = [batch for batch in train_iter]
On the last batch, I'll get error: Result of Slicing is an empty tensor

How should I fix it?

Thanks!

**Comment by jekbradbury** [comment]
I can't reproduce with `datasets.WikiText2` (I had to fix some unrelated bugs in that first, though). This code works for me after said unrelated fixes (PR coming momentarily):

```
from torchtext import data, datasets
TEXT = data.Field(sequential=True)
train, dev, test = datasets.WikiText2.splits(TEXT)
TEXT.build_vocab(train, dev, test)
train_iter, dev_iter, test_iter = data.BPTTIterator.splits(
    (train, dev, test), batch_size=20, bptt_len=35, device=-1, repeat=False)
for batch in train_iter:
    print(batch.text.size())
```
and prints sizes of 35 x 20 until the last batch of size 26 x 20.

Can you provide a snippet that's buggy?

**Comment by demiguo** [comment]
Hi Sure!

So In utils.py, we have:
`
def load_ptb(ptb_path='data.zip', ptb_dir='data', bptt_len=5, batch_size=1, gpu=False, 
        reuse=False, repeat=False, shuffle=True):
    print ("Loading Data")
    if (not reuse) or (not os.path.exists(ptb_dir)):
        f = zipfile.ZipFile(ptb_path, 'r')
        f.extractall('.')
        f.close()
    DEV = 0 if gpu else -1
    text_field = data.Field(lower=True, batch_first=True)
    train = datasets.LanguageModelingDataset(os.path.join(ptb_dir, 'train.txt'), text_field, newline_eos=False)
    val = datasets.LanguageModelingDataset(os.path.join(ptb_dir, 'valid.txt'), text_field, newline_eos=False)
    test = datasets.LanguageModelingDataset(os.path.join(ptb_dir, 'test.txt'), text_field, newline_eos=False)
    train_iter, val_iter, test_iter = data.BPTTIterator.splits((train, val, test), 
        batch_size=batch_size, bptt_len=bptt_len, device=DEV, repeat=repeat,
        shuffle=shuffle)
    text_field.build_vocab(train, vectors=torchtext.vocab.GloVe(name='6B', dim=50))
    return train_iter, val_iter, test_iter, text_field
`

And then in main.py function
Even `batches = [batch for batch in train_iter]` doesn't work

Thanks so much!

**Comment by jekbradbury** [comment]
I think I had an off-by-one error.

**Comment by jihunchoi** [comment]
The problem seems not to be fixed by #151; I am still getting the same error.
If I get it right, though I'm familiar with the language modeling task at all, the problem is in the `__len__` function.
Assume batch size is 4, BPTT length 10, and the length of the entire sequence 393042 (the numbers are from PTB-char validation split).
Note that the current implementation of `len(iterator)` returns 9827.
By padding and reshaping in line 234--238, `data` has size of `(98261, 4)`, thus `target` becomes an empty tensor.
I think the last batch should be ignored by correcting `len(iterator)` to return 9826.
Shortly I will send a PR to solve this.
